src/vault.py

```diff
@@ -48,14 +48,22 @@ def is_sealed(self) -> bool:
         """Returns whether Vault is sealed."""
         return self._client.sys.is_sealed()
 
-    def wait_for_unseal(self) -> None:
-        """Wait for Vault to be unsealed."""
-        elapsed_time = 0
-        while self.is_sealed() and elapsed_time < 30:
+    def wait_for_unseal(self, timeout: int = 30) -> None:
+        """Waits for Vault to be unsealed.
+
+        Expected to be called after attempting to unseal Vault.
+        If it times out, raises a TimeoutError.
+
+        Args:
+            timeout: Timeout in seconds.
+        """
+        initial_time = time.time()
+        while time.time() - initial_time < timeout:
+            if not self.is_sealed():
+                return
             logger.info("Vault is sealed, waiting for unseal")
-            time.sleep(5)
-            elapsed_time += 5
-        return
+            time.sleep(2)
```

Comment:
I like this pattern more (using time instead of counting). But there's a catch here:

Thinking about worst case scenario, pretend the scheduler doesn't get around to waking the process back up after 2 seconds. It's really busy and it takes 31 seconds. We will check the vault once right at the beginning, and sleep for 30 seconds, then raise a `TimeoutError`. Obviously this is an extreme situation to highlight a point, but it makes the problem clear I hope. We're waiting _at least_ 2 seconds in the end, for nothing (no additional work will be done afterwards).

Two ways to work around it come to mind:

1. Check if it's sealed before entering the loop, and sleep at the start of the loop instead of the end
2. Do one last check before you throw the error.

Reply:
Good catch, I added one more check.